include/pybind11/pybind11.h

```diff
@@ -803,59 +803,49 @@ class generic_type : public object {
-    using type = typename std::conditional<match, arg, typename recurse::type>::type;
-    using leftover = typename std::conditional<match, std::tuple<more...>,
-          decltype(std::tuple_cat(std::declval<std::tuple<arg>>(), std::declval<typename recurse::leftover>()))>::type;
+template <template<class> class Predicate, class... Ts>
+struct extract_first;
+
+template <template<class> class Predicate>
+struct extract_first<Predicate> {
+    constexpr static bool found = false;
+    using type = void;
 };
 
-template <typename T = void> struct extract_default { constexpr static bool found = false; using type = T; using leftover = std::tuple<>; };
-
-// Extract the first descendent of `base` from the parameter pack, or `void` if the parameter pack
-// has no descendent types.
-template <typename base, typename...> struct extract_subtype : extract_default<base> {};
-template <typename base, typename arg, typename... more>
-struct extract_subtype<base, arg, more...> : extract_if_match<std::is_base_of<base, arg>::value, arg,
-    extract_subtype<base, more...>, more...> {};
-// The same, but extracted from an std::tuple's types
-template <typename base, typename... Types> struct extract_subtype<base, std::tuple<Types...>> : extract_subtype<base, Types...> {};
-
-// Extracts the first base class of `type` from the parameter pack, or `void` if the parameter pack
-// has no ancestor types.
-template <typename type, typename...> struct extract_supertype : extract_default<void> {};
-template <typename type, typename arg, typename... more>
-struct extract_supertype<type, arg, more...> : extract_if_match<std::is_base_of<arg, type>::value, arg,
-    extract_supertype<type, more...>, more...> {};
-template <typename type, typename... Types> struct extract_supertype<type, std::tuple<Types...>> : extract_supertype<type, Types...> {};
-
-// Extracts a holder type for `base` from the parameter pack.  The holder_type must be either a unique_ptr, or
-// a type declared with PYBIND11_DECLARE_HOLDER_TYPE.  If none is found, yields a std::unique_ptr<base>.
-template <typename base, typename...> struct extract_holder_type : extract_default<std::unique_ptr<base>> {};
-template <typename base, typename arg, typename... more>
-struct extract_holder_type<base, arg, more...> : extract_if_match<is_holder_type<base, arg>::value, arg,
-    extract_holder_type<base, more...>, more...> {};
-// Extract from std::tuple types
-template <typename base, typename... Types> struct extract_holder_type<base, std::tuple<Types...>> : extract_holder_type<base, Types...> {};
+template <template<class> class Predicate, class T, class... Ts>
+struct extract_first<Predicate, T, Ts...> {
+    using type = typename std::conditional<
+        Predicate<T>::value,
+        T,
+        typename extract_first<Predicate, Ts...>::type
+    >::type;
+    constexpr static bool found = Predicate<T>::value || extract_first<Predicate, Ts...>::found;
```

Comment:
``found`` is probably not needed, can use ``std::is_same<...::type, void>`` below.